ln-gateway/src/ln.rs

```diff
@@ -66,3 +70,66 @@ impl LnRpc for Mutex<cln_rpc::ClnRpc> {
         }
     }
 }
+
+/// Control the behavoir of the LnRpc by exposing the same interface but attach configurations and
+/// methods to it.
+pub struct LnRpcAdapter {
+    // add more configuration
+    ln_client: Box<dyn LnRpc>,
+    fail_times: u8,
+    fail_invoice: Arc<Mutex<HashMap<String, AtomicU8>>>, //does this need to be in arc
+}
+
+impl LnRpcAdapter {
+    pub fn new(ln_client: Box<dyn LnRpc>, fail_times: u8) -> Self {
+        let fail_invoice = Arc::new(Mutex::new(HashMap::new()));
+
+        LnRpcAdapter {
+            ln_client,
+            fail_times,
+            fail_invoice,
+        }
+    }
+
+    /// Tell the LnRpc to fail the payment x-times until succeeding (or trying to succeed in the
+    /// case of real LnRpc
+    pub async fn fail_invoice(&self, invoice: String, times: u8) {
```

Comment:
Do we really need this mutex to be public? I'd have expected it to be an detail only accesible to impl LnRpc for the adapter

Reply:
I don't understand..all attributes of `LnRpcAdapter` are hidden. This function `fail_invoice` mutates it without showing that to the user.

Reply:
Oh I see how you apply this in the test case. Saw `fail_times` was a config in the last iteration of your PR and it seemed duplicate and unnecessary to have a public `fail_invoice` function.

Btw, I expected a simple fail x times before paying any given invoice.
This is way cooler because it supports fails on multiple invoices, independently :)🥇

Reply:
@okjodom we can also just add a `fail_any: Mutex<u8>` to it and a `pub fn fail_any(times: u8)` which just sets it..then just alter the `pay` impl analogous, thats the nice thing: things like that can just be plugged in like this

Reply:
ah ok understan, yes I changed quite a bit and even confused myself xd